testutils_test.go

```diff
@@ -115,6 +123,187 @@ func expectTraverse(t *testing.T, trav traverser, start, end string, count int)
+
+	curKeyIdx := 0
+	tree.Iterate(func(k, v []byte) bool {
+		nextMirrorKey := sortedMirrorKeys[curKeyIdx]
+		nextMirrorValue := mirror[nextMirrorKey]
+
+		require.Equal(t, []byte(nextMirrorKey), k)
+		require.Equal(t, []byte(nextMirrorValue), v)
+
+		curKeyIdx++
+		return false
+	})
+}
+
+func assertImmutableMirrorIterate(t *testing.T, tree *ImmutableTree, mirror map[string]string) {
+	sortedMirrorKeys := getSortedMirrorKeys(mirror)
+
+	curKeyIdx := 0
+	tree.Iterate(func(k, v []byte) bool {
+		nextMirrorKey := sortedMirrorKeys[curKeyIdx]
+		nextMirrorValue := mirror[nextMirrorKey]
+
+		require.Equal(t, []byte(nextMirrorKey), k)
+		require.Equal(t, []byte(nextMirrorValue), v)
+
+		curKeyIdx++
+		return false
+	})
+}
+
+func getSortedMirrorKeys(mirror map[string]string) []string {
+	sortedMirrorKeys := make([]string, 0, len(mirror))
+	for k := range mirror {
+		sortedMirrorKeys = append(sortedMirrorKeys, k)
+	}
+	sort.Strings(sortedMirrorKeys)
+	return sortedMirrorKeys
+}
+
+func getRandomizedTreeAndMirror(t *testing.T) (*MutableTree, map[string]string) {
+	const cacheSize = 100
+
+	tree, err := getTestTree(cacheSize)
+	require.NoError(t, err)
+
+	mirror := make(map[string]string)
+
+	randomizeTreeAndMirror(t, tree, mirror)
+	return tree, mirror
+}
+
+func randomizeTreeAndMirror(t *testing.T, tree *MutableTree, mirror map[string]string) {
+	if mirror == nil {
+		mirror = make(map[string]string)
+	}
+	const keyValLength = 5
+
+	numberOfSets := 1000
+	numberOfUpdates := numberOfSets / 4
+	numberOfRemovals := numberOfSets / 4
+
+	for numberOfSets > numberOfRemovals*3 {
+		key := randBytes(keyValLength)
+		value := randBytes(keyValLength)
+
+		isUpdated := tree.Set(key, value)
+		require.False(t, isUpdated)
+		mirror[string(key)] = string(value)
+
+		numberOfSets--
+	}
+
+	for numberOfSets+numberOfRemovals+numberOfUpdates > 0 {
+		randOp := rand.Intn(2)
+		if randOp == 0 && numberOfSets > 0 {
+
+			numberOfSets--
+
+			key := randBytes(keyValLength)
+			value := randBytes(keyValLength)
+
+			isUpdated := tree.Set(key, value)
+			require.False(t, isUpdated)
+			mirror[string(key)] = string(value)
+		} else if randOp == 1 && numberOfUpdates > 0 {
+
+			numberOfUpdates--
+
+			key := getRandomKeyFrom(mirror)
+			value := randBytes(keyValLength)
+
+			isUpdated := tree.Set([]byte(key), value)
+			require.True(t, isUpdated)
+			mirror[string(key)] = string(value)
+		} else if numberOfRemovals > 0 {
+
+			numberOfRemovals--
+
+			key := getRandomKeyFrom(mirror)
+
+			val, isRemoved := tree.Remove([]byte(key))
+			require.True(t, isRemoved)
+			require.NotNil(t, val)
+			delete(mirror, string(key))
+		}
+	}
+}
+
+func getRandomKeyFrom(mirror map[string]string) string {
+	keys := make([]string, 0, len(mirror))
+	for k := range mirror {
+		keys = append(keys, k)
+	}
+	key := keys[rand.Intn(len(keys))]
+	return key
+}
+
+func setupMirrorForIterator(t *testing.T, config *iteratorTestConfig, tree *MutableTree) [][]string {
+	mirror := make([][]string, 0)
```

Comment:
Went for the first one since performance here does not matter. However, other people could get confused in the future as for why we assumed a certain number